### PR TITLE
fix SendHeader/SetTrailer when use grpc-gateway standalone server (v2)

### PR DESCRIPTION
**Pull request by ghostiam**
Fix https://github.com/grpc-ecosystem/grpc-gateway/issues/517#issuecomment-684625645 when use gateway without gRPC.

### CODE REVIEW

**Comment by ghostiam** [comment]
@johanbrandhorst bazel failed by timeout. Very slow compiling...

File: protoc-gen-grpc-gateway/internal/gengateway/template.go [hunk 1]
```diff
@@ -595,13 +596,16 @@ func Register{{$svc.GetName}}{{$.RegisterFuncSuffix}}Server(ctx context.Context,
 		ctx, cancel := context.WithCancel(ctx)
 	{{- end }}
 		defer cancel()
+		var stream runtime.ServerTransportStream
```
Comment (johanbrandhorst) [inline protoc-gen-grpc-gateway/internal/gengateway/template.go:599]: We can update the comment on line 580 now.

File: runtime/context.go [hunk 2]
```diff
@@ -165,6 +166,59 @@ func ServerMetadataFromContext(ctx context.Context) (md ServerMetadata, ok bool)
 	return
 }
 
+// ServerTransportStream implement grpc.ServerTransportStream
```
Comment (johanbrandhorst) [inline runtime/context.go:169]:
```suggestion
// ServerTransportStream implements grpc.ServerTransportStream.
// It should only be used by the generated files to support grpc.SendHeader
// outside of gRPC server use.
```

File: runtime/context.go [hunk 3]
```diff
@@ -165,6 +166,59 @@ func ServerMetadataFromContext(ctx context.Context) (md ServerMetadata, ok bool)
...
+}
+
+// Method returns the method for the stream.
+func (s *ServerTransportStream) Method() string {
+	return ""
+}
+
+// Header returns the header metadata of the stream.
+func (s *ServerTransportStream) Header() metadata.MD {
+	c := s.header.Copy()
```
Comment (johanbrandhorst) [inline runtime/context.go:183]: Doesn't this need to be mutex protected?
```suggestion
	s.mu.Lock()
	defer s.mu.Unlock()
	c := s.header.Copy()
```

File: runtime/context.go [hunk 4]
```diff
@@ -165,6 +166,59 @@ func ServerMetadataFromContext(ctx context.Context) (md ServerMetadata, ok bool)
...
+}
+
+// SendHeader sets the header metadata.
+func (s *ServerTransportStream) SendHeader(md metadata.MD) error {
+	return s.SetHeader(md)
+}
+
+// Trailer returns the cached trailer metadata.
+func (s *ServerTransportStream) Trailer() metadata.MD {
+	c := s.trailer.Copy()
```
Comment (johanbrandhorst) [inline runtime/context.go:206]: Doesn't this need to be mutex protected?
```suggestion
	s.mu.Lock()
	defer s.mu.Unlock()
	c := s.trailer.Copy()
```

**Review by johanbrandhorst — CHANGES_REQUESTED** [review]

File: runtime/context.go (hunk 3 above)
Comment (ghostiam) [reply runtime/context.go:183]: Maybe, but the original library does not use a mutex in its implementation
https://github.com/grpc/grpc-go/blob/master/internal/transport/transport.go#L386

Comment (johanbrandhorst) [reply runtime/context.go:183]: Interesting, seems pretty specific to their design though. This design definitely needs it.

File: protoc-gen-grpc-gateway/internal/gengateway/template.go (hunk 1 above)
Comment (ghostiam) [reply protoc-gen-grpc-gateway/internal/gengateway/template.go:599]: I couldn't find any other examples, so I just removed the mention of grpc.SendHeader.

**Review by johanbrandhorst — APPROVED** [review]

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Pull request merged**